### PR TITLE
Add RSpec/ExampleWithoutDescription cop. Fixes #516

### LINKED ISSUE #516: Cop idea: it without description
RSpec allows for auto-generated descriptions, based on the expectation used. I personally don't like this style. But even if used, I've seen people put an empty string as description, instead of omitting it completely. So I propose cop that detect this and removes the empty string, as well as having an option to completely disable usage of it without a description

### PR DESCRIPTION
**Pull request by Darhazer**
Forbids usage of empty string as a description and also allows users to configure when it's allowed to have an example without any description 

-----------------

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) is passing.

If you have created a new cop:
* [x] Added the new cop to `config/default.yml`.
* [x] The cop includes examples of good and bad code.
* [x] You have tests for both code that should be reported and code that is good.

### CODE REVIEW

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+      #
+      # @example `EnforcedStyle: always_allow`
+      #   # bad
+      #   it('') { is_expected.to be_good }
+      #   it '' do
+      #     result = service.call
+      #     expect(result).to be(true)
+      #   end
+      #
+      #   it { is_expected.to be_good }
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: These are `# good`, right? (missing header)

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      #
+      # @example `EnforcedStyle: single_line_only`
+      #   # bad
+      #   it('') { is_expected.to be_good }
+      #   it do
+      #     result = service.call
+      #     expect(result).to be(true)
+      #   end
+      #
+      #   good
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: `#` missing.

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        def_node_matcher :example?, Examples::ALL.send_pattern
+        def_node_matcher :example_description, '(send nil? _ $(str $_))'
+
+        def on_send(node)
+          return unless example?(node)
+
+          check_example_without_description(node)
+
+          example_description(node) do |message_node, message|
+            return unless message.to_s.empty?
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: Just curious – what might `message` be besides a string?

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        private
+
+        def check_example_without_description(node)
+          return unless dissalow_empty_description?(node)
+
+          _send, _method, arg = *node
+          add_offense(node, message: MSG_ADD_DESCRIPTION) if arg.nil?
+        end
+
+        def dissalow_empty_description?(node)
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: Typo: dissalow -> disallow.

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+          return true if style == :single_line_only && node.parent.multiline?
+
+          style == :disallow
+        end
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: I’m always trying to come up with alternatives to `return <boolean logic>\n<other boolean logic>`. How about this:
```rb
def disallow_empty_description?(node)
  style == :disallow ||
    (style == :single_line_only && node.parent.multiline?)
end
```

File: lib/rubocop/cop/rspec/example_without_description.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            return unless message.to_s.empty?
+
+            add_offense(message_node, message: MSG_DEFAULT_ARGUMENT)
+          end
+        end
+
+        private
+
+        def check_example_without_description(node)
+          return unless dissalow_empty_description?(node)
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/example_without_description.rb]: I’d prefer the `arg.nil?` check earlier, before the `style` check here.

File: lib/rubocop/cop/rspec/example_without_description.rb (hunk 3 above)
Comment (Darhazer) [reply lib/rubocop/cop/rspec/example_without_description.rb]: Nothing else, I just have the habit of converting to the expected type

File: lib/rubocop/cop/rspec/example_without_description.rb (hunk 5 above)
Comment (Darhazer) [reply lib/rubocop/cop/rspec/example_without_description.rb]: I'm not fan either, but I also dislike mixing || and &&. Anyway, I'll use your version, as it looks cleaner indeed

**Comment by bquorning** [comment]
Please rebase on master to get a green build. A PR description would be nice, too :-)

**Pull request merged**